File: src/Application/Client/Java/Table.java
package com.scalien.scaliendb;

import java.math.BigInteger;
import java.util.List;
import java.util.Map;

/**
 * Table is a convenience class for encapsulating table related operations.
 * <p>
 * ScalienDB uses databases and tables to manage key value namespaces.
 * <p>
 * Example:
 * <pre>
 * db = client.getDatabase("testDatabase");
 * table = db.getTable("testTable");
 * // some sets
 * for (i = 0; i < 1000; i++) 
 *     table.set("foo" + i, "foo" + i);
 * client.submit();
 * for (i = 0; i < 1000; i++) 
 *     table.set("bar" + i, "bar" + i);
 * client.submit()
 * // some deletes
 * table.delete("foo0");
 * table.delete("foo10");
 * client.submit();
 * // count
 * System.out.println("number of keys starting with foo: " + table.count(new StringRangeParams().prefix("foo")));
 * // iterate
 * for (KeyValue&lt;String, String&gt; kv : table.getKeyValueIterator(new StringRangeParams().prefix("bar")))
 *     System.out.println(kv.getKey() + " => " + kv.getValue());
 * // truncate
 * table.truncate();
 * </pre>
 * @see Database#createTable(String) 
 * @see Database#createTable(String, Quorum)
 * @see Database#getTable(String) 
 * @see Database
 */
 public class Table
{
    Client client;
    private Database database;
    private long tableID;
    private String name;
    
    Table(Client client, Database database, long tableID, String name) throws SDBPException {
        this.client = client;
        this.database = database;
        this.tableID = tableID;
        this.name = name;
    }
    
    Client getClient() {
        return client;
    }
    
    long getTableID() {
        return tableID;
    }
    
    /** 
     * The database this table is in.
     */
    public Database getDatabase() {
        return database;
    }
    
    /**
     * The name of the table.
     */
    public String getName() {
        return name;
    }

    /**
     * Create a new instance of a query.
     */
    public Query createQuery() {
        return new Query(this);
    }

    /**
     * Rename the table.
     * @param newName The new name of the table.
     * @exception SDBPException 
     */
    public void renameTable(String newName) throws SDBPException {
        BigInteger biTableID = BigInteger.valueOf(tableID);
        int status = scaliendb_client.SDBP_RenameTable(client.cptr, biTableID, newName);
        client.checkResultStatus(status);    
    }

    /**
     * Delete the table. This means all key-values in the table are dropped.
     * @exception SDBPException 
     */
    public void deleteTable() throws SDBPException {
        BigInteger biTableID = BigInteger.valueOf(tableID);
        int status = scaliendb_client.SDBP_DeleteTable(client.cptr, biTableID);
        client.checkResultStatus(status);    
    }

    /**
     * Truncate the table.
     * @exception SDBPException 
     */
    public void truncateTable() throws SDBPException {
        BigInteger biTableID = BigInteger.valueOf(tableID);
        int status = scaliendb_client.SDBP_TruncateTable(client.cptr, biTableID);
        client.checkResultStatus(status);    
    }
 
    /**
     * Retrieve a value by key from the table. Returns <code>null</code> if not found.
     * @param key The key to look for.
     * @return The retrieved value.
     * @exception SDBPException 
     * @see #get(byte[]) 
     */
    public String get(String key) throws SDBPException {
        return client.get(tableID, key);
    }

    /**
     * Retrieve a value by key from the table. Returns <code>defval</code> if not found.
     * @param key The key to look for.
     * @param defval The default return value.
     * @return The retrieved value.
     * @exception SDBPException 
     * @see #get(byte[], byte[]) 
     */
    public String get(String key, String defval) throws SDBPException {
        String value = client.get(tableID, key);
        if (value != null)
            return value;
        else
            return defval;
    }

    /**
     * Retrieve a value by key from the table. Returns <code>null</code> if not found.
     * @param key The key to look for.
     * @return The retrieved value.
     * @exception SDBPException 
     * @see #get(String) 
     */
    public byte[] get(byte[] key) throws SDBPException {
        return client.get(tableID, key);
    }

    /**
     * Retrieve a value by key from the table. Returns <code>defval</code> if not found.
     * @param key The key to look for.
     * @param defval The default return value.
     * @return The retrieved value.
     * @exception SDBPException 
     * @see #get(String, String) 
     */
    public byte[] get(byte[] key, byte[] defval) throws SDBPException {
        byte[] value = client.get(tableID, key);
        if (value != null)
            return value;
        else
            return defval;
    }
    
    /**
     * Set a key-value in the table.
     * @param key The key.
     * @param value The value.
     * @exception SDBPException 
     * @see #set(byte[], byte[]) 
     */
    public void set(String key, String value) throws SDBPException {
        client.set(tableID, key, value);
    }

    /**
     * Set a key-value in the table.
     * @param key The key.
     * @param value The value.
     * @exception SDBPException 
     * @see #set(String, String) 
     */
    public void set(byte[] key, byte[] value) throws SDBPException {
        client.set(tableID, key, value);
    }
    
    /**
     * Delete a key-value pair by key in the table.
     * @param key The key.
     * @exception SDBPException 
     * @see #delete(byte[]) 
     */
    public void delete(String key) throws SDBPException {
        client.delete(tableID, key);
    }

    /**
     * Delete a key-value pair by key in the table.
     * @param key The key.
     * @exception SDBPException 
     * @see #delete(String) 
     */
    public void delete(byte[] key) throws SDBPException {
        client.delete(tableID, key);
    }
    
    /**
     * Return the number of matching keys in the table.
     * @param ps The filter parameters.
     * @return The number of matching keys in the table.
     * @exception SDBPException 
     * @see StringRangeParams 
     * @see #count(ByteRangeParams) 
     */
    public long count(StringRangeParams ps) throws SDBPException {
        return client.count(tableID, ps);
    }
    
    /**
     * Return the number of matching keys in the table.
     * @param ps The filter parameters.
     * @return The number of matching keys in the table.
     * @exception SDBPException 
     * @see ByteRangeParams 
     * @see #count(StringRangeParams) 
     */
    public long count(ByteRangeParams ps) throws SDBPException {
        return client.count(tableID, ps);
    }

    /**
     * Return an iterator that will return only keys.
     * @param ps The parameters of iteration, as a <a href="StringRangeParams.html">StringRangeParams</a>.
     * <p>
     * Example:
     * <pre>
     * db = client.getDatabase("testDatabase");
     * table = db.getTable("testTable");
     * for (String s : table.getKeyIterator(new StringRangeParams().prefix("foo")))
     *     System.out.println(s);
     * </pre>
     * @return The iterator.
     * @see StringRangeParams 
     * @see #getKeyIterator(ByteRangeParams) 
     * @see #getKeyValueIterator(StringRangeParams) 
     * @see #getKeyValueIterator(ByteRangeParams) 
     */
    public StringKeyIterator getKeyIterator(StringRangeParams ps) throws SDBPException {
        return new StringKeyIterator(this, ps);
    }

    /**
     * Return an iterator that will return only keys.
     * @param ps The parameters of iteration, as a <a href="ByteRangeParams.html">ByteRangeParams</a>.
     * @return The iterator.
     * @see ByteRangeParams 
     * @see #getKeyIterator(StringRangeParams) 
     * @see #getKeyValueIterator(StringRangeParams) 
     * @see #getKeyValueIterator(ByteRangeParams) 
     */
    public ByteKeyIterator getKeyIterator(ByteRangeParams ps) throws SDBPException {
        return new ByteKeyIterator(this, ps);
    }

    /**
     * Return an iterator that will return keys and values as a <code>KeyValue&lt;String, String&gt;</code>.
     * @param ps The parameters of iteration, as a <a href="StringRangeParams.html">StringRangeParams</a>.
     * <p>
     * Example:
     * <pre>
     * db = client.getDatabase("testDatabase");
     * table = db.getTable("testTable");
     * for (KeyValue&lt;String, String&gt; kv in table.getKeyValueIterator(new StringRangeParams().prefix("foo")))
     *     System.out.println(kv.getKey() + " => " + kv.getValue());
     * </pre>
     * @return The iterator.
     * @see StringRangeParams 
     * @see #getKeyValueIterator(ByteRangeParams) 
     * @see #getKeyIterator(StringRangeParams) 
     * @see #getKeyIterator(ByteRangeParams) 
     */
    public StringKeyValueIterator getKeyValueIterator(StringRangeParams ps) throws SDBPException {
        return new StringKeyValueIterator(this, ps);
    }

    /**
     * Return an iterator that will return keys and values as a <code>KeyValue&lt;byte[], byte[]&gt;</code>.
     * @param ps The parameters of iteration, as a <a href="ByteRangeParams.html">ByteRangeParams</a>.
     * @return The iterator.
     * @see ByteRangeParams 
     * @see #getKeyValueIterator(StringRangeParams) 
     * @see #getKeyIterator(StringRangeParams) 
     * @see #getKeyIterator(ByteRangeParams) 
     */
    public ByteKeyValueIterator getKeyValueIterator(ByteRangeParams ps) throws SDBPException {
        return new ByteKeyValueIterator(this, ps);
    }
}

File: src/Application/Client/CSharp/ScalienClient/Table.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Scalien
{
    /// <summary>
    /// Table is a convenience class for encapsulating table related operations.
    /// </summary>
    /// <remarks>
    /// <para>
    /// ScalienDB uses databases and tables to manage key value namespaces.
    /// </para>
    /// </remarks>
    /// <example><code>
    /// db = client.GetDatabase("testDatabase");
    /// table = db.GetTable("testTable");
    /// // some sets
    /// using (client.Begin())
    /// {
    ///     for (i = 0; i &lt; 1000; i++) 
    ///         table.Set("foo" + i, "foo" + i);
    /// }
    /// using (client.Begin())
    /// {
    ///     for (i = 0; i &lt; 1000; i++) 
    ///         table.Set("bar" + i, "bar" + i);
    /// }
    /// // some deletes
    /// table.Delete("foo0");
    /// table.Delete("foo10");
    /// client.Submit();
    /// // count
    /// System.Console.WriteLine("number of keys starting with foo: " + table.Count(new StringRangeParams().Prefix("foo")));
    /// // iterate
    /// foreach(KeyValuePair&lt;string, string&gt; kv in table.GetKeyValueIterator(new StringRangeParams().Prefix("bar")))
    ///     System.Console.WriteLine(kv.Key + " => " + kv.Value);
    /// // truncate
    /// table.Truncate();
    /// </code></example>
    /// <seealso cref="Scalien.Database.CreateTable(string)"/>
    /// <seealso cref="Scalien.Database.CreateTable(string, Quorum)"/>
    /// <seealso cref="Scalien.Database.GetTable(string)"/>
    /// <seealso cref="Database"/>
    public class Table
    {
        private Client client;
        private Database database;
        private ulong tableID;
        private string name;

        #region Properties

        internal Client Client
        {
            get
            {
                return client;
            }
        }

        internal ulong TableID
        {
            get
            {
                return tableID;
            }
        }

        /// <summary>
        /// The database this table is in.
        /// </summary>
        public Database Database
        {
            get
            {
                return database;
            }
        }

        /// <summary>
        /// The name of the table.
        /// </summary>
        public string Name
        {
            get
            {
                return name;
            }
        }

        #endregion

        #region Constructors, destructors

        internal Table(Client client, Database database, ulong tableID, string name)
        {
            this.client = client;
            this.database = database;
            this.tableID = tableID;
            this.name = name;
        }

        #endregion

        #region Table management

        /// <summary>
        /// Rename the table.
        /// </summary>
        /// <param name="newName">The new name of the table.</param>
        /// <exception cref="SDBPException"/>
        public void RenameTable(string newName)
        {
            int status = scaliendb_client.SDBP_RenameTable(client.cptr, tableID, newName);
            client.CheckResultStatus(status);
        }

        /// <summary>
        /// Delete the table.
        /// </summary>
        /// <exception cref="SDBPException"/>
        public void DeleteTable()
        {
            int status = scaliendb_client.SDBP_DeleteTable(client.cptr, tableID);
            client.CheckResultStatus(status);
        }

        /// <summary>
        /// Truncate the table. This means all key-values in the table are dropped.
        /// </summary>
        /// <exception cref="SDBPException"/>
        public void TruncateTable()
        {
            int status = scaliendb_client.SDBP_TruncateTable(client.cptr, tableID);
            client.CheckResultStatus(status);
        }

        #endregion

        #region Data commands

        /// <summary>
        /// Retrieve a value by key from the table. Returns null if not found.
        /// </summary>
        /// <param name="key">The key to look for.</param>
        /// <returns>The retrieved value.</returns>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Get(byte[])"/>
        public string Get(string key)
        {
            return client.Get(tableID, key);
        }

        /// <summary>
        /// Retrieve a value by key from the table. Returns null if not found.
        /// </summary>
        /// <param name="key">The key to look for.</param>
        /// <returns>The retrieved value.</returns>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Get(string)"/>
        public byte[] Get(byte[] key)
        {
            return client.Get(tableID, key);
        }

        /// <summary>
        /// Set a key-value in the table.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <param name="value">The value.</param>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Set(byte[], byte[])"/>
        public void Set(string key, string value)
        {
            client.Set(tableID, key, value);
        }

        /// <summary>
        /// Set a key-value in the table.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <param name="value">The value.</param>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Set(string, string)"/>
        public void Set(byte[] key, byte[] value)
        {
            client.Set(tableID, key, value);
        }

        /// <summary>
        /// Delete a key-value pair by key in the table.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Delete(byte[])"/>
        public void Delete(string key)
        {
            client.Delete(tableID, key);
        }

        /// <summary>
        /// Delete a key-value pair by key in the table.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Delete(string)"/>
        public void Delete(byte[] key)
        {
            client.Delete(tableID, key);
        }

        /// <summary>
        /// Return the number of matching keys in the table.
        /// </summary>
        /// <param name="ps">The filter parameters.</param>
        /// <returns>The number of matching keys in the table.</returns>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Scalien.StringRangeParams"/>
        /// <seealso cref="Count(ByteRangeParams)"/>
        public ulong Count(StringRangeParams ps)
        {
            return client.Count(tableID, ps);
        }

        /// <summary>
        /// Return the number of matching keys in the table.
        /// </summary>
        /// <param name="ps">The filter parameters.</param>
        /// <returns>The number of matching keys in the table.</returns>
        /// <exception cref="SDBPException"/>
        /// <seealso cref="Scalien.ByteRangeParams"/>
        /// <seealso cref="Count(StringRangeParams)"/>
        public ulong Count(ByteRangeParams ps)
        {
            return client.Count(tableID, ps);
        }

        #endregion

        #region Iterators

        /// <summary>
        /// Return an iterator that will return only keys.
        /// </summary>
        /// <param name="ps">The parameters of iteration, as a <see cref="Scalien.StringRangeParams"/>.</param>
        /// <returns>The iterator.</returns>
        /// <example><code>
        /// db = client.GetDatabase("testDatabase");
        /// table = db.GetTable("testTable");
        /// foreach (string s in table.GetKeyIterator(new StringRangeParams().Prefix("foo")))
        ///     System.Console.WriteLine(s);
        /// </code></example>
        /// <seealso cref="Scalien.StringRangeParams"/>
        /// <seealso cref="GetKeyIterator(ByteRangeParams)"/>
        /// <seealso cref="GetKeyValueIterator(StringRangeParams)"/>
        /// <seealso cref="GetKeyValueIterator(ByteRangeParams)"/>
        public StringKeyIterator GetKeyIterator(StringRangeParams ps)
        {
            return new StringKeyIterator(this, ps);
        }

        /// <summary>
        /// Return an iterator that will return only keys.
        /// </summary>
        /// <param name="ps">The parameters of iteration, as a <see cref="Scalien.ByteRangeParams"/>.</param>
        /// <returns>The iterator.</returns>
        /// <seealso cref="Scalien.ByteRangeParams"/>
        /// <seealso cref="GetKeyIterator(ByteRangeParams)"/>
        /// <seealso cref="GetKeyValueIterator(StringRangeParams)"/>
        /// <seealso cref="GetKeyValueIterator(ByteRangeParams)"/>
        public ByteKeyIterator GetKeyIterator(ByteRangeParams ps)
        {
            return new ByteKeyIterator(this, ps);
        }

        /// <summary>
        /// Return an iterator that will return keys and values as a <see cref="System.Collections.Generic.KeyValuePair{T, T}"/>.
        /// </summary>
        /// <param name="ps">The parameters of iteration, as a <see cref="Scalien.StringRangeParams"/>.</param>
        /// <returns>The iterator.</returns>
        /// <example><code>
        /// db = client.GetDatabase("testDatabase");
        /// table = db.GetTable("testTable");
        /// foreach (KeyValuePair&lt;string, string&gt; kv in table.GetKeyValueIterator(new StringRangeParams().Prefix("foo")))
        ///     System.Console.WriteLine(kv.Key + " => " + kv.Value);
        /// </code></example>
        /// <seealso cref="System.Collections.Generic.KeyValuePair{T, T}"/>
        /// <seealso cref="Scalien.StringRangeParams"/>
        /// <seealso cref="GetKeyValueIterator(ByteRangeParams)"/>
        /// <seealso cref="GetKeyIterator(StringRangeParams)"/>
        /// <seealso cref="GetKeyIterator(ByteRangeParams)"/>
        public StringKeyValueIterator GetKeyValueIterator(StringRangeParams ps)
        {
            return new StringKeyValueIterator(this, ps);
        }

        /// <summary>
        /// Return an iterator that will return keys and values as a <see cref="System.Collections.Generic.KeyValuePair{T, T}"/>.
        /// </summary>
        /// <param name="ps">The parameters of iteration, as a <see cref="Scalien.ByteRangeParams"/>.</param>
        /// <returns>The iterator.</returns>
        /// <seealso cref="System.Collections.Generic.KeyValuePair{T, T}"/>
        /// <seealso cref="Scalien.ByteRangeParams"/>
        /// <seealso cref="GetKeyValueIterator(StringRangeParams)"/>
        /// <seealso cref="GetKeyIterator(StringRangeParams)"/>
        /// <seealso cref="GetKeyIterator(ByteRangeParams)"/>
        public ByteKeyValueIterator GetKeyValueIterator(ByteRangeParams ps)
        {
            return new ByteKeyValueIterator(this, ps);
        }

        #endregion

        #region Retrieve helpers: Sequence, Submitter

        /// <summary>
        /// Retrieve a <see cref="Scalien.Sequence"/> backed by a key-value in this table.
        /// </summary>
        /// <param name="key">The key backing the sequence.</param>
        /// <returns>The corresponding <see cref="Scalien.Sequence"/> object.</returns>
        /// <seealso cref="GetSequence(byte[])"/>
        /// <seealso cref="Scalien.Sequence"/>
        public Sequence GetSequence(string key)
        {
            return new Sequence(this.client, this.tableID, key);
        }

        /// <summary>
        /// Retrieve a <see cref="Scalien.Sequence"/> backed by a key-value in this table.
        /// </summary>
        /// <param name="key">The key backing the sequence.</param>
        /// <returns>The corresponding <see cref="Scalien.Sequence"/> object.</returns>
        /// <seealso cref="GetSequence(string)"/>
        /// <seealso cref="Scalien.Sequence"/>
        public Sequence GetSequence(byte[] key)
        {
            return new Sequence(this.client, this.tableID, key);
        }

        #endregion
    }
}
